ewojjowe/graphql-api
# Refactor codes and add authentication on request header

### Summary of Changes
 - Refactor and clean up the code
 - Update user schema and model
 - Apply authentication middleware
 - Apply authentication middleware
 - Add login resolver
 - Check authentication on resolvers

### Acceptance Criteria
 - Should accept an authenticated request
 - Should reject the unauthenticated request

### How to Test
 - checkout branch
 - run server

 - in postman checking, navigate to `http://localhost:8080/graphql` using post request method
 - to generate token add the following query below on request body and hit send: 
 ```
note: request body should be raw and in json format

{
	"query": "query { login(loginInput: {email:\"fred@test.com\", password:\"akosiya\"}) {token}}"
}
```
 - copy the token and replace the request body to follwing query:
```
{
	"query": "mutation {createEvent(eventInput: {title:\"Should Work no. 3\", description:\"akosiya\", price: 2323.55, date:\"2019-06-04T20:29:37.669Z\"}) {_id title creator {_id email}}}"
}
```
 - add Authorization on header with the value format - `Bearer <token>`, sample below, then hit send:
```
Bearer eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJ1c2VySWQiOiI1Y2Y2YzBlMWY5MjgxNjRhMmU5YjEyNWUiLCJlbWFpbCI6ImZyZWRAdGVzdC5jb20iLCJpYXQiOjE1NTk2Nzk5NTYsImV4cCI6MTU1OTY4MzU1Nn0.s_nnS0RJAY3yKWDXtQ1fA0kMHc9hS7deVKJ16T4xWys
```
 - should successfuly created event

 - on browser checking, navigate to `http://localhost:8080/graphql and use the following queries to test:

```
// mutation query for creating a booking
mutation {
  bookEvent(eventId: "5cf6c114bf95bd4a38ef75f2") {
    _id
    event {
      title
      creator {
        _id
      }
    }
  }
}

// mutation query for cancel booking
mutation {
  cancelBooking(bookingId: "5cf6c12dbf95bd4a38ef75f4") {
    title
    description
  }
}

// queries all booking
query {
  bookings {
    _id
    createdAt
    updatedAt
    event {
      _id
      title
      description
      price
      date
    }
    user {
      email
      password
    }
  }
}
```
Should reject all querues above.
